File: kotlin/src/test/kotlin/aether/content/ChunkBitmapFixtureTest.kt
// SPDX-License-Identifier: MIT
package aether.content

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.DynamicTest.dynamicTest
import org.junit.jupiter.api.TestFactory
import java.io.File
import java.nio.file.Paths
import java.util.Base64

/**
 * Cross-language ChunkBitmap wire-format fixture verifier — Kotlin runner.
 *
 * Reads fixtures/content/chunk_bitmap_vectors.json and verifies that this
 * implementation produces bit-identical bitsets and JSON payloads for each
 * pinned test vector.
 */

// ── Model ──────────────────────────────────────────────────────────────────────

@Serializable
private data class ChunkBitmapVector(
    val name: String,
    val description: String,
    @SerialName("root_hash")         val rootHash: String,
    @SerialName("chunk_count")       val chunkCount: Int,
    @SerialName("have_indices")      val haveIndices: List<Int>,
    @SerialName("have_bitset_hex")   val haveBitsetHex: String,
    @SerialName("have_bitset_base64") val haveBitsetBase64: String,
    val generation: Long,
    @SerialName("expected_json")     val expectedJson: String,
)

class ChunkBitmapFixtureTest {

    // ── Inline BitsetCodec (spec compliance — does not import from main src) ───

    private fun inlineEncode(chunkCount: Int, indices: List<Int>): ByteArray {
        if (chunkCount <= 0) return ByteArray(0)
        val bytes = ByteArray((chunkCount + 7) / 8)
        for (i in indices) {
            bytes[i shr 3] = (bytes[i shr 3].toInt() or (1 shl (i and 7))).toByte()
        }
        return bytes
    }

    private fun inlineDecode(bitset: ByteArray, chunkCount: Int): List<Int> {
        val result = mutableListOf<Int>()
        val limit = minOf(chunkCount, bitset.size * 8)
        for (i in 0 until limit) {
            if ((bitset[i shr 3].toInt() and 0xFF) and (1 shl (i and 7)) != 0) result.add(i)
        }
        return result
    }

    private fun inlineMarshal(
        rootHash: String,
        chunkCount: Int,
        haveBitset: ByteArray,
        generation: Long,
    ): String {
        val b64 = Base64.getEncoder().encodeToString(haveBitset)
        return """{"root_hash":"$rootHash","chunk_count":$chunkCount,"have_bitset":"$b64","generation":$generation}"""
    }

    // ── Fixture loader ─────────────────────────────────────────────────────────

    private val jsonParser = Json { ignoreUnknownKeys = true }

    private fun loadVectors(): List<ChunkBitmapVector> {
        var dir = File(Paths.get("").toAbsolutePath().toString())
        repeat(12) {
            val candidate = File(dir, "fixtures/content/chunk_bitmap_vectors.json")
            if (candidate.exists()) {
                return jsonParser.decodeFromString<List<ChunkBitmapVector>>(candidate.readText())
            }
            dir = dir.parentFile ?: return@repeat
        }
        throw IllegalStateException("Could not locate fixtures/content/chunk_bitmap_vectors.json")
    }

    private val vectors by lazy { loadVectors() }

    // ── Tests ──────────────────────────────────────────────────────────────────

    @TestFactory
    fun `encode produces correct bitset`(): List<DynamicTest> = vectors.map { v ->
        dynamicTest(v.name) {
            val bitset = inlineEncode(v.chunkCount, v.haveIndices)
            val hex = bitset.joinToString("") { "%02x".format(it) }
            assertEquals(v.haveBitsetHex.lowercase(), hex)
            assertEquals(v.haveBitsetBase64, Base64.getEncoder().encodeToString(bitset))
        }
    }

    @TestFactory
    fun `decode recovers correct indices`(): List<DynamicTest> = vectors.map { v ->
        dynamicTest(v.name) {
            val bitset = Base64.getDecoder().decode(v.haveBitsetBase64)
            val recovered = inlineDecode(bitset, v.chunkCount)
            assertEquals(v.haveIndices.sorted(), recovered.sorted())
        }
    }

    @TestFactory
    fun `JSON serialization matches expected`(): List<DynamicTest> = vectors.map { v ->
        dynamicTest(v.name) {
            val bitset = inlineEncode(v.chunkCount, v.haveIndices)
            val actual = inlineMarshal(v.rootHash, v.chunkCount, bitset, v.generation)
            assertEquals(v.expectedJson, actual)
        }
    }

    @TestFactory
    fun `bitset length is ceil div 8`(): List<DynamicTest> = vectors.map { v ->
        dynamicTest(v.name) {
            val bitset = inlineEncode(v.chunkCount, v.haveIndices)
            assertEquals((v.chunkCount + 7) / 8, bitset.size)
        }
    }

    @TestFactory
    fun `trailing bits are zero`(): List<DynamicTest> = vectors.map { v ->
        dynamicTest(v.name) {
            val bitset = inlineEncode(v.chunkCount, v.haveIndices)
            if (bitset.isEmpty()) return@dynamicTest
            val trailing = v.chunkCount % 8
            if (trailing == 0) return@dynamicTest
            val last = bitset.last().toInt() and 0xFF
            val validMask = (1 shl trailing) - 1
            assertEquals(0, last and validMask.inv().and(0xFF))
        }
    }
}

File: typescript/tests/chunk_bitmap_fixture.test.ts
// SPDX-License-Identifier: MIT
/**
 * Cross-language ChunkBitmap wire-format fixture verifier — TypeScript runner.
 *
 * Reads fixtures/content/chunk_bitmap_vectors.json and verifies that this
 * implementation produces bit-identical bitsets and JSON payloads for each
 * pinned test vector.
 */

import { describe, it } from "node:test";
import { strict as assert } from "node:assert";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, resolve } from "node:path";

// ── Inline implementations (spec compliance — no src/ imports) ────────────────

function bitsetEncode(chunkCount: number, haveIndices: readonly number[]): Uint8Array {
  if (chunkCount <= 0) return new Uint8Array(0);
  const bytes = new Uint8Array(Math.ceil(chunkCount / 8));
  for (const i of haveIndices) bytes[i >> 3] |= 1 << (i & 7);
  return bytes;
}

function bitsetDecode(bitset: Uint8Array, chunkCount: number): number[] {
  const result: number[] = [];
  const limit = Math.min(chunkCount, bitset.length * 8);
  for (let i = 0; i < limit; i++)
    if (bitset[i >> 3] & (1 << (i & 7))) result.push(i);
  return result;
}

function marshalJson(rootHash: string, chunkCount: number, haveBitset: Uint8Array, generation: number): string {
  const b64 = Buffer.from(haveBitset).toString("base64");
  return `{"root_hash":${JSON.stringify(rootHash)},"chunk_count":${chunkCount},"have_bitset":${JSON.stringify(b64)},"generation":${generation}}`;
}

// ── Fixture loader ─────────────────────────────────────────────────────────────

interface ChunkBitmapVector {
  name: string;
  description: string;
  root_hash: string;
  chunk_count: number;
  have_indices: number[];
  have_bitset_hex: string;
  have_bitset_base64: string;
  generation: number;
  expected_json: string;
}

function findFixtures(): string {
  let dir = dirname(fileURLToPath(import.meta.url));
  for (let i = 0; i < 10; i++) {
    const candidate = resolve(dir, "fixtures", "content", "chunk_bitmap_vectors.json");
    try { readFileSync(candidate); return candidate; } catch {}
    const parent = dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  throw new Error("Could not locate fixtures/content/chunk_bitmap_vectors.json");
}

const vectors: ChunkBitmapVector[] = JSON.parse(readFileSync(findFixtures(), "utf8"));

// ── Tests ──────────────────────────────────────────────────────────────────────

describe("ChunkBitmap — cross-language fixtures", () => {
  describe("encode produces correct bitset", () => {
    for (const v of vectors) {
      it(v.name, () => {
        const bitset = bitsetEncode(v.chunk_count, v.have_indices);
        assert.equal(Buffer.from(bitset).toString("hex"), v.have_bitset_hex.toLowerCase());
        assert.equal(Buffer.from(bitset).toString("base64"), v.have_bitset_base64);
      });
    }
  });

  describe("decode recovers correct indices", () => {
    for (const v of vectors) {
      it(v.name, () => {
        const bitset = new Uint8Array(Buffer.from(v.have_bitset_base64, "base64"));
        const recovered = bitsetDecode(bitset, v.chunk_count);
        assert.deepEqual([...recovered].sort((a,b)=>a-b), [...v.have_indices].sort((a,b)=>a-b));
      });
    }
  });

  describe("JSON serialization matches expected", () => {
    for (const v of vectors) {
      it(v.name, () => {
        const bitset = bitsetEncode(v.chunk_count, v.have_indices);
        const actual = marshalJson(v.root_hash, v.chunk_count, bitset, v.generation);
        assert.equal(actual, v.expected_json);
      });
    }
  });

  describe("bitset length is ceil(chunkCount/8)", () => {
    for (const v of vectors) {
      it(v.name, () => {
        const bitset = bitsetEncode(v.chunk_count, v.have_indices);
        assert.equal(bitset.length, Math.ceil(v.chunk_count / 8));
      });
    }
  });

  describe("trailing bits are zero", () => {
    for (const v of vectors) {
      it(v.name, () => {
        const bitset = bitsetEncode(v.chunk_count, v.have_indices);
        if (bitset.length === 0) return;
        const trailingBits = v.chunk_count % 8;
        if (trailingBits === 0) return;
        const lastByte = bitset[bitset.length - 1];
        const validMask = (1 << trailingBits) - 1;
        assert.equal(lastByte & ~validMask, 0);
      });
    }
  });
});
